src/util/indexrange.h

```diff
@@ -0,0 +1,175 @@
+#ifndef MIXXX_INDEXRANGE_H
+#define MIXXX_INDEXRANGE_H
+
+#include <QtDebug>
+
+#include <iosfwd>
+#include <utility>
+
+#include "util/assert.h"
+#include "util/types.h"
+
+
+namespace mixxx {
+
+// A half-open, directed range of indices with limited mutability.
+class IndexRange final: private std::pair<SINT, SINT> {
+    typedef std::pair<SINT, SINT> Super;
+
+  public:
+    using Super::swap;
+
+    IndexRange()
+        : Super(0, 0) {
+        DEBUG_ASSERT(empty());
+    }
+
+    static IndexRange between(SINT head, SINT tail) {
+        return IndexRange(head, tail);
+    }
+    static IndexRange forward(SINT head, SINT length) {
+        DEBUG_ASSERT(length >= 0);
+        return IndexRange(head, head + length);
+    }
+    static IndexRange backward(SINT head, SINT length) {
+        DEBUG_ASSERT(length >= 0);
+        return IndexRange(head, head - length);
+    }
+
+    // The near boundary (inclusive)
+    SINT head() const {
+        return first;
+    }
+    // The opposite boundary (exclusive)
+    SINT tail() const {
+        return second;
+    }
+
+    bool empty() const {
+        return head() == tail();
+    }
+
+    enum class Orientation {
+        Empty,
+        Forward,
+        Backward,
+    };
+    Orientation orientation() const {
+        if (empty()) {
+            return Orientation::Empty; // undefined
+        } else {
+            if (head() < tail()) {
+                return Orientation::Forward;
+            } else {
+                return Orientation::Backward;
+            }
+        }
+    }
+
+    SINT length() const {
+        return (head() <= tail()) ? (tail() - head()) : (head() - tail());
+    }
+
+    // Clamps index by this range including both head() and tail()
+    // boundaries.
+    SINT clamp(SINT index) const {
+        if (head() <= tail()) {
+            return std::max(head(), std::min(tail(), index));
+        } else {
+            return std::min(head(), std::max(tail(), index));
+        }
+    }
+
+    bool contains(SINT index) const {
+        if (head() <= tail()) {
+            return (head() <= index) && (index < tail());
+        } else {
+            return (head() >= index) && (index > tail());
+        }
+    }
+
+    // Splits this range into two adjacent parts by slicing off
+    // and returning a range of given length and same direction
+    // from the head side. The given length must not exceed
+    // the length of this range.
```

Comment:
Those mutable operations sacrifice safety for usability. In languages like Rust or Scala that support pattern matching for variable bindings I would have done it the way you proposed.